### PR TITLE
investigate escaping jumping to definition (and other features) with spaces

### LINKED ISSUE #403: Jump to definition handling spaces in URI inconsistently
## Description

Jump to definition feature doesn't seem to handle spaces in base/parent path correctly. Jump to-feature behaves as if trying to perform a global jump using the .lsp_symlink, even though the target is in same directory than jump source. Getting similar results both in Windows 10 and Ubuntu 18.04:

image

## Reproduce
Not entirely sure if prerequisite in reproducing the issue, but I have [jupytext ](https://github.com/mwouts/jupytext) installed.

1. Start jupyterlab server in directory, which has spaces in the name (might be reproducible with just space in anywhere in path)
2. Create a target file (extension.py in my example) with function definition
3. Create and open another python file in notebook mode and import function
4. Jump to function definition 

While debugging the issue I pinpointed the culprit to:
https://github.com/krassowski/jupyterlab-lsp/blob/fbc9d1c2f67842352acbbcb276347c6e65d51d29/packages/jupyterlab-lsp/src/utils.ts#L134
where child and parent:
`child = "file:///c:/Users/avaissi/Documents/My own robot/.virtual_documents/tasks/extension.py"`
`parent = "file:///c:/Users/avaissi/Documents/My%20own%20robot"`
thus failing the `child.startsWith(parent)` comparison.

When tried to wrap the `parent || server_root_uri()` with `decodeURI()` -> `parent = decodeURI(parent || server_root_uri());` things worked nicely.
## Expected behavior

I'm expecting the jump to complete to a file/location  within same directory.
And furthermore, I'm expecting the URI to compare spaces (and possibly other special characters) either encoded or decoded.

## Context
- Windows 10, Ubuntu 18.04
- Chrome: Version 86.0.4240.183 (Official Build) (64-bit)
- JupyterLab version: 2.2.6
- jupyterlab-lsp: 2.0.8

<details><summary>Required: installed server extensions</summary>
<pre>
jupyter_lsp 0.9.2 ok
jupyterlab 2.2.6 ok
jupytext 1.5.0 ok
</pre>
</details>

<details><summary>Required: installed lab extensions</summary>
<pre>
@jupyter-widgets/jupyterlab-manager v2.0.0 enabled  ok
@krassowski/jupyterlab-lsp v2.0.8 enabled  ok
@robocorp/theme-robocodelab-extension v2.3.2 enabled  ok
jupyterlab-jupytext v1.2.1 enabled  ok
</pre>
</details>

<details><summary>Browser Output (recommended for all interface issues)</summary>
<pre>
Will jump to the first of suggested locations: Array(1)0: {uri: "file:///c:/Users/avaissi/Documents/My%20own%20robot/.virtual_documents/tasks/extension.py", range: {…}}length: 1__proto__: Array(0)
jump_to.js:77 Jumping to external file: file:///c:/Users/avaissi/Documents/My own robot/.virtual_documents/tasks/extension.py
jump_to.js:78 Jump target (source location): Object
api/contents/c%3A/Users/avaissi/Documents/My%20own%20robot/.virtual_documents/tasks/extension.py?content=0&1604934607968:1 Failed to load resource: the server responded with a status of 404 (Not Found)
jump_to.js:103 Error: c:/Users/avaissi/Documents/My own robot/.virtual_documents/tasks/extension.py is outside root contents directory
    at Function.create (serverconnection.js:106)
    at async Drive.get (index.js:496)
    at async jump_to_CMJumpToDefinition.handle_jump (jump_to.js:96)
    at async Object.execute (jump_to.js:137)
handle_jump @ jump_to.js:103
notebook.js:75 LSP: waiting for .lsp_symlink//c:Users/avaissi/Documents/My own robot/.virtual_documents/tasks/extension.py to fully load
api/contents/.lsp_symlink/c%3AUsers/avaissi/Documents/My%20own%20robot/.virtual_documents/tasks/extension.py?type=notebook&content=1&1604934607990:1 Failed to load resource: the server responded with a status of 404 (Not Found)
</pre>
</details>

### PR DESCRIPTION
**Pull request by bollwyvl**
## References

- fixes #403

## Code changes

- [x] atest
  - [x] demonstrate a reproducible failure (bb39801)
  - [x] rework some globals to be set by `Variables.robot` (can be overridden from CLI)
  - [x] move more lab config into JSON, launch without `shell=True`
- [x] `@krassowski/jupyterlab-lsp`
  - [x] remove a number of `(decode|encode)URI` calls
  - [x] always `encodeURI` the document path
- [x] `@krassowski/jupyterlab_go_to_definition`
  - [x] rename `IGlobalJump.uri` to `IGlobalJump.contents_path`
  - [x] add some more information to the docstrings
  - [x] sweep for more references to URIs that are actually contents paths
- [x] consider any additional ways to prevent this in the future
  - e.g. more normalization at boundaries?
  - find/build/contribute a better URI library (#410)

## User-facing changes

Should reliably Jump between files in folders with spaces and special characters, including when `jupyter lab` is launched from directories containing spaces, especially on windows.

## Backwards-incompatible changes

- `@krassowski/jupyterlab_go_to_definition` 
  - `IGlobalJump.uri` renamed to `IGlobalJump.contents_path`

## Chores

- [x] linted
- [x] tested
- [ ] documented
- [x] changelog entry

### CODE REVIEW

**Comment by bollwyvl** [comment]
Ha: failed to even _start_ the acceptance tests on windows, so haven't yet reproduced (even though it's failing).

We're _generally_ very careful about paths, but the way we start jupyter under test is quite kludgy... I'll look at doing more of it with a json file which is less sensitive to paths/escaping.

**Comment by avaissi** [comment]
Anything here I could help with? Maybe to investigate the issue more for root cause, if not already found

**Comment by bollwyvl** [comment]
Unfortunately, it's a bit in the weeds! I have a plan, it's just going to
take some time. Spaces, permissions, and unicode should be table stakes for
acceptance testing anything, but it's easy to forget. Thanks again for the
report!

File: scripts/atest.py [hunk 1]
```diff
@@ -80,7 +83,7 @@ def atest(attempt, extra_args):
         if previous.exists():
             extra_args += ["--rerunfailed", str(previous)]
 
-    out_dir = OUT / stem
+    out_dir = OUT / (stem + GH_403_BAD_PATH)
```
Comment (krassowski) [inline scripts/atest.py:86]: I wonder if the path addition should not be pushed further down to the actual robot tests. I think that we already have one notebook with space in the name, probably just not the one testing jump functions. Though having this as a top dir would be convenient indeed.

**Comment by bollwyvl** [comment]
Yeah the issue is "spaces in the rootUri break everything". Including
starting the test server, at present!

File: scripts/atest.py (hunk 1 above)
Comment (bollwyvl) [reply scripts/atest.py]: Indeed! I was kind of tired. 

Moved it (and a number of other paths) down into `Variables.robot`... they can be overridden with:

```bash
python -m scripts.atest --variable "NOTEBOOK DIR NAME:🐶 🐶 🐶"
```
would yield:
```bash
atest/output/linux_37_1/home/🐶 🐶 🐶/
```

**Comment by bollwyvl** [comment]
gah... the formatting of the flake8 messages now leads with the code, e.g. `F821 undefined name`. Will try with `title*=[<the diagnostic>]`, as who knows...

File: atest/07_Configuration.robot [hunk 2]
```diff
@@ -40,8 +40,8 @@ LaTeX
 *** Keywords ***
 Settings Should Change Editor Diagnostics
     [Arguments]    ${language}    ${file}    ${server}    ${settings}    ${before}    ${after}    ${save command}=${EMPTY}    ${needs reload}=${False}
-    ${before diagnostic} =    Set Variable    ${CSS DIAGNOSTIC}\[title^="${before}"]
-    ${after diagnostic} =    Set Variable    ${CSS DIAGNOSTIC}\[title^="${after}"]
+    ${before diagnostic} =    Set Variable    ${CSS DIAGNOSTIC}\[title*="${before}"]
+    ${after diagnostic} =    Set Variable    ${CSS DIAGNOSTIC}\[title*="${after}"]
```
Comment (bollwyvl) [inline atest/07_Configuration.robot:44]: @krassowski this could be cherry-picked to `master` to get CI working...

**Comment by bollwyvl** [comment]
Well, still didn't get to reproducing #403 on windows (or at least hasn't finished), but otherwise good progress...

**Comment by bollwyvl** [comment]
Great news! I can reproduce the error!

selenium-screenshot-1

Hopefully we'll get just that fail for all of the tests, and I can add the `decodeURI` part!

**Comment by bollwyvl** [comment]
Actually had a look deeper: turns out the `IGlobalJump.uri` is not a URI at all, and has always been the more "humane" Jupyter contents path. This in turn gets URL encoded when it actually hits the `ContentsManager`, as HTTP _is_ ASCII, like it or lump it.

A few changes throughout were required to normalize this behavior (86f5abe), and make the naming more accurate (89102dd, breaking change), and linting (2e43577) but it should _hopefully_ now be more robust overall, and address some other odd issues I had seen but hadn't narrowed down to this root cause.

On #278 I briefly toyed with using an existing URI library, but ran into some issues... it may be worth revisiting, either:
- adopting somebody's,
- getting one into lab/lumino
- at worst rolling our own 
  - which can already handle things like `URI.toContentsPath()`, etc. 

At any rate, whatever we use, we should probably _never_ treat a URI as a plain-old-`string`, which would prvent these kinds of issues, as well as double encode, etc. 

Going to let a full local clean build run while CI is chugging, and might not address any more this evening... good times!

**Comment by bollwyvl** [comment]
Local run looks good, and 89102dd has test wins (aside from lint) on all the platforms:
https://github.com/krassowski/jupyterlab-lsp/runs/1404308642

Taking off WIP, ready for review!

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,23 @@
 ## CHANGELOG
 
+### `@krassowski/jupyterlab-lsp 2.0.9` (???)
+
+- bug fixes
+
+  - handles characters that need escaping (spaces, non-ASCII characters) more
+    robustly in files and folder names ([#403])
+
+[#403]: https://github.com/krassowski/jupyterlab-lsp/pulls/403
```
Comment (krassowski) [inline CHANGELOG.md:10]: 406?

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,23 @@
...
+
+### `@krassowski/jupyterlab_go_to_definition 2.0.0` (???)
+
+- features
+
+  - breaking change: renames `uri` to `contents_path` to help avoid programmer issues
+    with characters requiring URI encoding ([#406])
+
+[#406]: https://github.com/krassowski/jupyterlab-lsp/pulls/406
```
Comment (krassowski) [inline CHANGELOG.md:19]:
```suggestion
[#406]: https://github.com/krassowski/jupyterlab-lsp/pull/406
```

File: CHANGELOG.md (hunk 3 above)
Comment (krassowski) [reply CHANGELOG.md:10]: or 403 issue?

Comment (krassowski) [reply CHANGELOG.md:10]:
```suggestion
[#403]: https://github.com/krassowski/jupyterlab-lsp/issues/403
```

**Comment by krassowski** [comment]
Ok, looks great will merge once CI passes. Thank you!

**Comment by bollwyvl** [comment]
Yeah, if there is an issue, and it's just little old me that did the work,
I just put the issue number, instead of The Work... I'm sure whatever state
is in now (if you pushed)  is fine! Very good though, to give shout outs
new contributor prs, though!

**Comment by BugApe** [comment]
This problem also occurs when the filename contains "=", such as "NeuralTaskonomy_axis=1.ipynb", and renaming the variable will fail.
image

**Comment by krassowski** [comment]
What language server are you using?

**Comment by BugApe** [comment]
> What language server are you using?

python

**Comment by krassowski** [comment]
@BugApe, which one:
- pyls
- pylsp
- jedi-ls
- pyright

? Could you open a new issue following bug report template?

**Pull request merged**